module/VuFindSearch/src/VuFindSearch/Backend/WorldCat2/Connector.php

```diff
@@ -0,0 +1,211 @@
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+
+namespace VuFindSearch\Backend\WorldCat2;
+
+use Exception;
+use Laminas\Http\Client\Exception\RuntimeException as ExceptionRuntimeException;
+use Laminas\Http\Exception\InvalidArgumentException;
+use Laminas\Http\Exception\RuntimeException;
+use Laminas\Http\Response;
+use Laminas\Log\LoggerAwareInterface;
+use Laminas\Session\Container;
+use League\OAuth2\Client\Provider\Exception\IdentityProviderException;
+use League\OAuth2\Client\Provider\GenericProvider;
+use VuFind\Log\LoggerAwareTrait;
+use VuFindSearch\ParamBag;
+
+/**
+ * Class for accessing OCLC WorldCat search API v2.
+ *
+ * @category VuFind
+ * @package  WorldCat
+ * @author   Demian Katz <demian.katz@villanova.edu>
+ * @license  http://opensource.org/licenses/gpl-2.0.php GNU General Public License
+ * @link     https://vufind.org/wiki/development Wiki
+ */
+class Connector implements LoggerAwareInterface
+{
+    use LoggerAwareTrait;
+
+    /**
+     * API base URL
+     *
+     * @var string
+     */
+    protected $baseUrl;
+
+    /**
+     * Constructor
+     *
+     * @param \Laminas\Http\Client $client       An HTTP client object
+     * @param GenericProvider      $authProvider OAuth2 provider
+     * @param Container            $session      Session container for persisting data
+     * @param array                $options      Additional config settings
+     */
+    public function __construct(
+        protected \Laminas\Http\Client $client,
+        protected GenericProvider $authProvider,
+        protected Container $session,
+        protected array $options = []
+    ) {
+        if (empty($options['base_url'])) {
+            throw new \Exception('base_url setting is required');
+        }
+        $this->baseUrl = $options['base_url'];
+    }
+
+    /**
+     * Get an OAuth2 token.
+     *
+     * @return string
+     * @throws IdentityProviderException
+     */
+    protected function getToken(): string
+    {
+        return $this->authProvider->getAccessToken('client_credentials', ['scope' => 'wcapi'])->getToken();
+    }
+
+    /**
+     * Make an API call.
+     *
+     * @param string $path Path to query (including GET parameters)
+     *
+     * @return Response
+     * @throws IdentityProviderException
+     * @throws InvalidArgumentException
+     * @throws RuntimeException
+     * @throws ExceptionRuntimeException
+     */
+    protected function makeApiCall(string $path): Response
+    {
+        if (!isset($this->session->token)) {
+            $this->session->token = $this->getToken();
+        }
+        $headers = ['Authorization: Bearer ' . $this->session->token];
+        $this->client->setHeaders($headers);
+        $this->client->setUri($this->baseUrl . $path);
+        $this->debug($path);
+        $response = $this->client->send();
+        // If authorization failed, the token may be expired; re-request and try again:
```

Comment:
'Authorization' would be 403 not 401 (authentication), but actually the line below should perhaps catch both?  People may have a variety of OCLC API keys, some authorized for different APIs so would get a 403.

Reply:
I think a 403 would indicate a valid token was being sent to an inappropriate service, and in that context, I don't think requesting a refresh of the token would make any difference. You're right that my comment is misleading, though; I have just fixed it.

Reply:
Token expired will return a 401. Token for the wrong set of scopes a 403.

Reply:
Ok I agree, there's no point in retrying after a 403.